### ISSUE TITLE
POSKeyError when getting session

### ISSUE THREAD

**Issue by mauritsvanrees** [zopefoundation/Products.Sessions#4]
Original report to Plone: https://community.plone.org/t/sessions-and-poskeyerrors/5939
Second report to zodb-dev: https://groups.google.com/forum/#!topic/zodb/RPJKEFfAHxI

The text below is mostly a copy of my latest mail to zodb-dev, with a bit more introduction. And it looks nicer when formatted. :-)

Summary: when getting a session from the request, I sometimes get a POSKeyError from the tempstorage. I can't reliably reproduce it, but these should roughly be the prerequisites:

- Do several requests in parallel, handled by different threads, like when the browser loads the html and then several resources: css, javascript, images.
- In every request, access the session
- Maybe also write to the session.
- Maybe it only triggers when a BTree gets reshuffled.

Environment:
- This is for a new site that is in development, so I only tested it locally on my Mac.  One Zope instance, no ZEO setup. Two threads.
- Using Python 2.7.14, ZODB 5.3.0, Zope2 2.13.26, tempstorage 4.0.1, BTrees 4.4.1, Plone 5.1.
- The main possible problem seems to be in `Products.Transience.Transience`, which has not changed between Zope 2 and Zope 4, except whitespace, removal of an icon, and splitting it out into the `Products.Sessions` package. So I thought I would report it here in this issue tracker.

Today I got a traceback, including debugging info, from setting the `Z_TOC_DEBUG` environment variable, which makes `Products.Transience` print debugging info.

At first some debugging info where all still seems fine:

```
2018-03-07 11:39:12 INFO Transience 123145425707008 1520419152.26 _move_item: keys for ts 1520418980 (bucket 4534646400)-- ['89172937J1x4MDNceGM1XHhlOVx4MTBceGUyXHhkOFx4MGYiJw']
2018-03-07 11:39:12 INFO Transience 123145429913600 1520419152.26 _move_item: bucket for ts 1520419020 is 4542801248
2018-03-07 11:39:12 INFO Transience 123145425707008 1520419152.26 _move_item: found_ts is 1520418980
2018-03-07 11:39:12 INFO Transience 123145429913600 1520419152.27 _move_item: keys for ts 1520419020 (bucket 4542801248)-- []
2018-03-07 11:39:12 INFO Transience 123145425707008 1520419152.27 _move_item: current_ts (1520419140) != found_ts (1520418980), moving to current
2018-03-07 11:39:12 INFO Transience 123145429913600 1520419152.27 _move_item: bucket for ts 1520419000 is 4542801392
2018-03-07 11:39:12 INFO Transience 123145425707008 1520419152.27 _move_item: keys for found_ts 1520418980 (bucket 4534646400): ['89172937J1x4MDNceGM1XHhlOVx4MTBceGUyXHhkOFx4MGYiJw']
2018-03-07 11:39:12 INFO Transience 123145429913600 1520419152.27 _move_item: keys for ts 1520419000 (bucket 4542801392)-- []
2018-03-07 11:39:12 INFO Transience 123145425707008 1520419152.27 _move_item: copied item 89172937J1x4MDNceGM1XHhlOVx4MTBceGUyXHhkOFx4MGYiJw from 1520418980 to 1520419140 (bucket 4544108016)
2018-03-07 11:39:12 INFO Transience 123145429913600 1520419152.27 _move_item: bucket for ts 1520418980 is 4542801824
2018-03-07 11:39:12 INFO Transience 123145425707008 1520419152.27 _move_item: deleted item 89172937J1x4MDNceGM1XHhlOVx4MTBceGUyXHhkOFx4MGYiJw from ts 1520418980 (bucket 4534646400)
2018-03-07 11:39:12 INFO Transience 123145429913600 1520419152.27 _move_item: keys for ts 1520418980 (bucket 4542801824)-- ['89172937J1x4MDNceGM1XHhlOVx4MTBceGUyXHhkOFx4MGYiJw']
2018-03-07 11:39:12 INFO Transience 123145425707008 1520419152.27 _move_item: returning 89172937J1x4MDNceGM1XHhlOVx4MTBceGUyXHhkOFx4MGYiJw from current_ts 1520419140 
2018-03-07 11:39:12 INFO Transience 123145429913600 1520419152.27 _move_item: found_ts is 1520418980
2018-03-07 11:39:12 INFO Transience 123145429913600 1520419152.27 _move_item: current_ts (1520419140) != found_ts (1520418980), moving to current
2018-03-07 11:39:12 INFO Transience 123145429913600 1520419152.27 _move_item: keys for found_ts 1520418980 (bucket 4542801824): ['89172937J1x4MDNceGM1XHhlOVx4MTBceGUyXHhkOFx4MGYiJw']
2018-03-07 11:39:12 INFO Transience 123145429913600 1520419152.27 _move_item: copied item 89172937J1x4MDNceGM1XHhlOVx4MTBceGUyXHhkOFx4MGYiJw from 1520418980 to 1520419140 (bucket 4542800384)
2018-03-07 11:39:12 INFO Transience 123145429913600 1520419152.27 _move_item: deleted item 89172937J1x4MDNceGM1XHhlOVx4MTBceGUyXHhkOFx4MGYiJw from ts 1520418980 (bucket 4542801824)
```

And then, presumably in a different request, it starts going wrong:

```
2018-03-07 11:39:12 INFO ZPublisher.Conflict ReadConflictError at /Plone/++plone++static/++unique++2018-03-02 16:36:34.751840/plone-legacy-compiled.css: '\x00\x00\x00\x00\x00\x00\x00+' (1 conflicts (0 unresolved) since startup at Wed Mar  7 11:36:11 2018)
2018-03-07 11:39:12 INFO Transience 123145425707008 1520419152.27 new_or_existing called with 89172937J1x4MDNceGM1XHhlOVx4MTBceGUyXHhkOFx4MGYiJw
2018-03-07 11:39:12 INFO Transience 123145425707008 1520419152.27 get: called with key 89172937J1x4MDNceGM1XHhlOVx4MTBceGUyXHhkOFx4MGYiJw, default []
2018-03-07 11:39:12 INFO Transience 123145425707008 1520419152.27 _finalize: lock acquired successfully
2018-03-07 11:39:12 INFO Transience 123145429913600 1520419152.58 new_or_existing called with 89172937J1x4MDNceGM1XHhlOVx4MTBceGUyXHhkOFx4MGYiJw
2018-03-07 11:39:12 INFO Transience 123145429913600 1520419152.58 get: called with key 89172937J1x4MDNceGM1XHhlOVx4MTBceGUyXHhkOFx4MGYiJw, default []
2018-03-07 11:39:12 INFO Transience 123145429913600 1520419152.58 _finalize: could not acquire lock, returning
2018-03-07 11:39:12 INFO Transience 123145429913600 1520419152.58 _replentish: attempting optional replentish (lock acquired)
2018-03-07 11:39:12 INFO Transience 123145429913600 1520419152.58 roll: low: 76020957, high: 76020977: lost with 76020967 (optional replentish)
2018-03-07 11:39:12 INFO Transience 123145429913600 1520419152.58 roll: low: 0, high: 5: won with 0 (gc)
2018-03-07 11:39:12 INFO Transience 123145429913600 1520419152.58 _gc:  (1520419140 -1520418960) > 160.0, gc invoked
2018-03-07 11:39:12 INFO Transience 123145429913600 1520419152.58 _do_gc_work: entering
2018-03-07 11:39:12 ERROR ZODB.Connection Couldn't load state for Products.Transience.Transience.Increaser 0x09
Traceback (most recent call last):
  File "/Users/maurits/shared-eggs/cp27m/ZODB-5.3.0-py2.7.egg/ZODB/Connection.py", line 796, in setstate
    p, serial = self._storage.load(oid)
  File "/Users/maurits/shared-eggs/cp27m/ZODB-5.3.0-py2.7.egg/ZODB/mvccadapter.py", line 143, in load
    r = self._storage.loadBefore(oid, self._start)
  File "/Users/maurits/shared-eggs/cp27m/tempstorage-4.0.1-py2.7.egg/tempstorage/TemporaryStorage.py", line 187, in loadBefore
    raise POSException.POSKeyError(oid)
POSKeyError: 0x09
```
And a couple more for good measure:

```
2018-03-07 11:39:12 ERROR ZODB.Connection Couldn't load state for Products.Transience.Transience.Increaser 0x09
Traceback (most recent call last):
  File "/Users/maurits/shared-eggs/cp27m/ZODB-5.3.0-py2.7.egg/ZODB/Connection.py", line 796, in setstate
    p, serial = self._storage.load(oid)
  File "/Users/maurits/shared-eggs/cp27m/ZODB-5.3.0-py2.7.egg/ZODB/mvccadapter.py", line 143, in load
    r = self._storage.loadBefore(oid, self._start)
  File "/Users/maurits/shared-eggs/cp27m/tempstorage-4.0.1-py2.7.egg/tempstorage/TemporaryStorage.py", line 187, in loadBefore
    raise POSException.POSKeyError(oid)
POSKeyError: 0x09
2018-03-07 11:39:12 ERROR client.myplugin KeyError getting SESSION for request http://localhost:8080/Plone/++theme++barceloneta/less/barceloneta-compiled.css
Traceback (most recent call last):
  File "/Users/maurits/clients/client/project/src/client.myplugin/src/client/myplugin/plugins.py", line 31, in get_session
    return request.SESSION
  File "/Users/maurits/shared-eggs/cp27m/Zope2-2.13.26-py2.7.egg/ZPublisher/HTTPRequest.py", line 1380, in __getattr__
    v = self.get(key, default, returnTaints=returnTaints)
  File "/Users/maurits/shared-eggs/cp27m/Zope2-2.13.26-py2.7.egg/ZPublisher/HTTPRequest.py", line 1337, in get
    v = v()
  File "/Users/maurits/shared-eggs/cp27m/Zope2-2.13.26-py2.7.egg/Products/Sessions/SessionDataManager.py", line 101, in getSessionData
    return self._getSessionDataObject(key)
  File "/Users/maurits/shared-eggs/cp27m/Zope2-2.13.26-py2.7.egg/Products/Sessions/SessionDataManager.py", line 188, in _getSessionDataObject
    ob = container.new_or_existing(key)
  File "/Users/maurits/shared-eggs/cp27m/Zope2-2.13.26-py2.7.egg/Products/Transience/Transience.py", line 839, in new_or_existing
    item = self.get(key, _marker)
  File "/Users/maurits/shared-eggs/cp27m/Zope2-2.13.26-py2.7.egg/Products/Transience/Transience.py", line 495, in get
    item = self._move_item(k, current_ts, default)
  File "/Users/maurits/shared-eggs/cp27m/Zope2-2.13.26-py2.7.egg/Products/Transience/Transience.py", line 285, in _move_item
    self._housekeep(current_ts)
  File "/Users/maurits/shared-eggs/cp27m/Zope2-2.13.26-py2.7.egg/Products/Transience/Transience.py", line 934, in _housekeep
    self._finalize(now)
  File "/Users/maurits/shared-eggs/cp27m/Zope2-2.13.26-py2.7.egg/Products/Transience/Transience.py", line 551, in _finalize
    last_finalized = self._last_finalized_timeslice()
  File "/Users/maurits/shared-eggs/cp27m/Zope2-2.13.26-py2.7.egg/Products/Transience/Transience.py", line 1071, in __call__
    return self.value
  File "/Users/maurits/shared-eggs/cp27m/ZODB-5.3.0-py2.7.egg/ZODB/Connection.py", line 796, in setstate
    p, serial = self._storage.load(oid)
  File "/Users/maurits/shared-eggs/cp27m/ZODB-5.3.0-py2.7.egg/ZODB/mvccadapter.py", line 143, in load
    r = self._storage.loadBefore(oid, self._start)
  File "/Users/maurits/shared-eggs/cp27m/tempstorage-4.0.1-py2.7.egg/tempstorage/TemporaryStorage.py", line 187, in loadBefore
    raise POSException.POSKeyError(oid)
POSKeyError: 0x09
2018-03-07 11:39:12 WARNING client.myplugin SESSION is None in extractCredentials.
```

So the error happens in `Products.Transience` in `_do_gc_work`.  At least that is the last line before the first error:

    2018-03-07 11:39:12 INFO Transience 123145429913600 1520419152.58 _do_gc_work: entering

That method looks like this:

```
    def _do_gc_work(self, now):
        # this is only separated from _gc for readability; it should
        # generally not be called by anything but _gc

        # we garbage collect any buckets that have already been run
        # through finalization
        DEBUG and TLOG('_do_gc_work: entering')

        max_ts = self._last_finalized_timeslice()

        DEBUG and TLOG('_do_gc_work: max_ts is %s' % max_ts)
	...
```

Since the second DEBUG line is not printed, it must go wrong in the line between, when calling `self._last_finalized_timeslice`.
That is an instance of a very simple Persistent class in `Transience.py`:

```
class Increaser(Persistent):
    """
    A persistent object representing a typically increasing integer that
    has conflict resolution which uses the greatest integer out of the three
    available states.
    """
    def __init__(self, v):
        self.value = v

    def set(self, v):
        self.value = v

    def __getstate__(self):
        return self.value

    def __setstate__(self, v):
        self.value = v

    def __call__(self):
        return self.value

    def _p_resolveConflict(self, old, state1, state2):
        return max(old, state1, state2)
```

Does this give anyone an idea of what to fix where?

Note: my current plan is to avoid using a session for this, and instead store what I need in a cookie.  That should also avoid possible problems in a setup with multiple zeoclients.

**Comment by icemac** [comment]
This issue looks like the one I recorded in zopefoundation/tempstorage#8. Maybe they have the same reason behind them.

**Comment by cekk** [comment]
I've seen that there are some updates here: https://github.com/zopefoundation/Zope/issues/650

I still have these problems reading session data from request. Like @mauritsvanrees after some reads, the session manager broke and i can't read infos from it anymore unless i restart the instance.

@mauritsvanrees do you have some newer informations since last reply?

**Comment by mauritsvanrees** [comment]
No newer information, sorry.

**Comment by ghiodor** [comment]
I'm not sure it is the same issue, but with current Zope 4.1 I get easily POSKey errors in the temporary storage.  The object /temp_folder/session_data is actually reported as "Broken object". I can open it and view its properties. In fact, the very first time i viewed it and then browsed to the container folder I got the error below. I also got the error in this line of a PythonScript session = request.SESSION.

Products.Sessions seems prety much unusable. Unfortunately I don't have yet a reproduceable case.

```
Request URL = http://localhost:8787/temp_folder/session_data/manage_container

Traceback (innermost last):

    Module ZPublisher.WSGIPublisher, line 155, in transaction_pubevents
    Module ZPublisher.WSGIPublisher, line 337, in publish_module
    Module ZPublisher.WSGIPublisher, line 247, in publish
    Module ZPublisher.mapply, line 85, in mapply
    Module ZPublisher.WSGIPublisher, line 61, in call_object
    Module App.special_dtml, line 99, in __call__
    Module DocumentTemplate.DT_String, line 527, in __call__
    Module DocumentTemplate._DocumentTemplate, line 145, in render_blocks
    Module DocumentTemplate._DocumentTemplate, line 215, in render_blocks_
    Module DocumentTemplate._DocumentTemplate, line 376, in __getitem__
    Module DocumentTemplate._DocumentTemplate, line 400, in getitem
    Module Products.Transience.Transience, line 983, in housekeep
    Module Products.Transience.Transience, line 988, in _housekeep
    Module Products.Transience.Transience, line 784, in _gc
    Module Products.Transience.Transience, line 1137, in __call__
    Module ZODB.Connection, line 791, in setstate
    Module ZODB.mvccadapter, line 143, in load
    Module tempstorage.TemporaryStorage, line 187, in loadBefore

ZODB.POSException.POSKeyError: 0x09
```

**Comment by d-maurer** [comment]
ghiodor wrote at 2019-7-20 04:31 -0700:
>I'm not sure it is the same issue, but with current Zope 4.1 I get easily POSKey errors in the temporary storage.  The object /temp_folder/session_data is actually reported as "Broken object". I can open it and view its properties. In fact, the very first time i viewed it and then browsed to the container folder I got the error below. I also got the error in this line of a PythonScript session = request.SESSION.
>
>Products.Sessions seems prety much unusable. Unfortunately I don't have yet a reproduceable case.
>
>```
>Request URL = http://localhost:8787/temp_folder/session_data/manage_container
>
>Traceback (innermost last):
> ...
>    Module Products.Transience.Transience, line 1137, in __call__
>    Module ZODB.Connection, line 791, in setstate
>    Module ZODB.mvccadapter, line 143, in load
>    Module tempstorage.TemporaryStorage, line 187, in loadBefore
>
>ZODB.POSException.POSKeyError: 0x09
>```

Almost surely, the problem is not in `Products.Sessions`
but in `tempstorage`.

I strongly recommend not to use `tempstorage`.
`tempstorage` has two known weaknesses:

 1. to save RAM, it maintains very limitied history. As a consequence,
    ZODB's MVCC (= "Multi Version Concurrency Control") often does
    not find the required historical data and responds with
    a `POSKeyError`.

    This is particularly grave, if sessions are frequently
    (e.g. in every request) accessed. Note that a reading
    session access may internally cause a write (due to session
    timeout handling) and then modifiy the ZODB.

 2. to reduce the conflict risk, it reduces the strength
    of the transaction isolation. This could lead to
    inconsistencies.


If you want to use Zope sessions, put them into a "normal" storage.
Even then, avoid to access the session in every request (used cookies
for those use cases).

**Comment by ghiodor** [comment]
Thanks for your feedback @d-maurer. I think this feedback deserves to be in the Products.Session readme file since it currently recommends temporarystorage. It would be good to put there also how to setup "normal" storage. It's not obvious to me. 

But please note that I was using Sessions in Zope2 without problem (maybe 100 bytes per user < 1000). With zope4.1+py3.8, Session breaks very easily (even at the first use of a new instance - not reproducible).

**Comment by cekk** [comment]
Thanks @d-maurer i had PoskeyError reading the session on each request, so that could be the cause.
As @ghiodor i didn't knew about temp storage problems, so i am gonna think twice about it in the future projects.

Btw i already switched to cookies in that project ;)

**Comment by d-maurer** [comment]
ghiodor wrote at 2019-7-21 03:00 -0700:
>Thanks for your feedback @d-maurer. I think this feedback deserves to be in the Products.Session readme file since it currently recommends temporarystorage. It would be good to put there also how to setup "normal" storage. It's not obvious to me.

The easiest way is to put the Zope sessions into the "main" storage.
To do this, you have two options:
 1. delete `temp_folder` and replace it by a normal folder. Put
    a "Transient Object Container" with id `session_data` into the new folder.

 2. reconfigure the `session_data_manager` to point to a
    `Transient Object Container` of your choice.


If you want to give the Zope sessions their own storage, change
your Zope configuration file and replace the `zodb_db` definition
for `temporary` by your own definition (it may look similar
to the `zodb_db` definition for the `main` storage; but you
may keep the `mount-point` from `temporary`).


Note that Plone's CSRF protection expects sessions to have their
own storage (and potentially even that it is a `tempstorage`;
this would be a bug in my opinion).

**Comment by dataflake** [comment]
@mauritsvanrees If you think the fixes in release 4.6 are enough to document the fact that tempstorage is just broken feel free ahead and close the issue.

**Comment by mauritsvanrees** [comment]
@dataflake Agreed. Thanks for your help!
